extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/cache/UpdateCounter.java

```diff
@@ -34,14 +39,39 @@ void update()
 
   void awaitTotalUpdates(int totalUpdates) throws InterruptedException
   {
+    totalUpdates &= MAX_PHASE;
     int currentUpdates = phaser.getPhase();
-    while (totalUpdates - currentUpdates > 0) { // overflow-aware
+    checkNotTerminated(currentUpdates);
+    while (comparePhases(totalUpdates, currentUpdates) > 0) {
       currentUpdates = phaser.awaitAdvanceInterruptibly(currentUpdates);
+      checkNotTerminated(currentUpdates);
+    }
+  }
+
+  private static int comparePhases(int phase1, int phase2)
+  {
+    int diff = (phase1 - phase2) & MAX_PHASE;
```

Comment:
Phase is always between 0 and Integer.MAX_VALUE. The essence of this bug is that before, the code implied that phase goes though the whole integer range, i. e. after Integer.MAX_VALUE the next phase is Integer.MIN_VALUE. Actually the next phase after Integer.MAX_VALUE is 0.